uPortSDK/Classes/EthrDID/EthrDIDResolver.swift

```diff
@@ -0,0 +1,334 @@
+            if baseEvent is EthereumDIDRegistry.Events.DIDDelegateChanged.Arguments {
+                let event = baseEvent as! EthereumDIDRegistry.Events.DIDDelegateChanged.Arguments
+                let delegateTypeOptional = String( data: event.delegatetype.value, encoding: .utf8 )
+                let delegate = event.delegate.encodeUnpadded().withoutHexPrefix.withHexPrefix
+                let key = "DIDDelegateChanged-\(delegateTypeOptional ?? "<unknown-delegatetype>")-\(delegate)"
+                let validTo = Int64( event.validto.value )
+                
+                if now <= validTo {
+                    delegateCount += 1
+                    
+                    guard let delegateType = delegateTypeOptional else {
+                        print( "invalid delegate type" )
+                        continue
+                    }
+                    
+                    switch delegateType {
+                    case DelegateType.Secp256k1SignatureAuthentication2018.rawValue, self.sigAuth:
+                        authEntries[ key ] = AuthenticationEntry(type: .Secp256k1SignatureAuthentication2018, publicKey: "\(normalizedDid)#delegate-\(delegateCount)")
+                    case DelegateType.Secp256k1VerificationKey2018.rawValue, self.veriKey:
+                        pkEntries[ key ] = PublicKeyEntry(id: "\(normalizedDid)#delegate-\(delegateCount)", type: .Secp256k1VerificationKey2018, owner: normalizedDid, ethereumAddress: delegate)
+                    default:
+                        print( "unknown delegateType -> \(delegateType)")
+                    }
+                }
+            }
+            
+            if baseEvent is EthereumDIDRegistry.Events.DIDAttributeChanged.Arguments {
+                let event = baseEvent as! EthereumDIDRegistry.Events.DIDAttributeChanged.Arguments
+                let validTo = Int64( event.validto.value )
+                if now <= validTo {
+                    let nameOptional = String( data: event.name.value, encoding: .utf8 )
+                    
+                    guard let name = nameOptional else {
+                        print( "invalid DIDAttributeChanged event name" )
+                        continue
+                    }
+                    
+                    let key = "DIDAttributeChanged-\(name)-\(event.value.value.hexEncodedString())"
+                    let ( section, algorithm, rawType, encoding ) = try self.destructuredAttributeChanged(eventName: name)
+                    let type = try self.parseType( algorithm: algorithm, rawType: rawType )
+                    switch section {
+                    case "pub":
+                        delegateCount += 1
+                        let publicKeyOwner = "\(normalizedDid)#delegate-\(delegateCount)"
+                        var pk = PublicKeyEntry(id: publicKeyOwner, type: type, owner: normalizedDid)
+                        switch encoding {
+                        case "", "null", "hex":
```

Comment:
There's an error here:

"Expression pattern of type 'String' cannot match values of type 'String?'"
I'm not sure if this is coming from my older xcode (9.2) or is a genuine swift error

The error is fixed when switching on a non-optional variable.

Reply:
I pushed a small commit fixing this